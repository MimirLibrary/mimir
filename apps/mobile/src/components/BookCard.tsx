import React, { FC } from 'react';
import { Image, StyleProp, StyleSheet, View, ViewStyle } from 'react-native';
import { AppText } from './AppText';
import { TextVariant } from '../config/textVariants';
import { Color, shadowVariants } from '../config/designTokens';

const DummyImage = require('../assets/unsplash_aZ_MmSmAcjg.png');

enum BookStatus {
  OVERDUE,
  ReturnTill,
}

interface IBookCard {
  style: StyleProp<ViewStyle>;
}

export const BookCard: FC<IBookCard> = ({ style }) => {
  const bookStatus = BookStatus.OVERDUE;

  const renderStatus = () => {
    const overdue = BookStatus.OVERDUE === bookStatus;

    return (
      <View
        style={[
          styles.bookStatusContainer,
          overdue && styles.overdueBookContainer,
        ]}
      >
        <AppText
          variant={TextVariant.CAPTION_M}
          color={overdue ? Color.TEXT_ERROR : Color.ACCENT_BLUE_100}
          text={overdue ? 'Overdue' : 'Return till: 26.04.2022'}
        />
      </View>
    );
  };

  return (
    <View style={[styles.container, style]}>
      <Image style={styles.image} resizeMode={'stretch'} source={DummyImage} />
      <View style={styles.infoContainer}>
        <View style={styles.textWrapper}>
          <AppText variant={TextVariant.CAPTION_R} text={'Psychology'} />
          <AppText
            variant={TextVariant.BODY_M}
            style={styles.title}
            text={'The Psychology of Money Second part. New version'}
          />
          <AppText
            variant={TextVariant.BODY_L}
            style={styles.description}
            text={'Morgan Housel'}
          />
        </View>
        {renderStatus()}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    padding: 10,
    borderRadius: 10,
    marginBottom: 8,
    backgroundColor: Color.WHITE,
    ...shadowVariants.shadowXL,
  },
  image: { width: 80, height: 130 },
  textWrapper: { width: '85%' },
  infoContainer: { marginLeft: 16, justifyContent: 'space-between' },
  description: { marginTop: 10 },
  title: { marginTop: 8 },
  bookStatusContainer: {
    alignSelf: 'flex-start',
    flexDirection: 'row',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 4,
    padding: 4,
    paddingHorizontal: 10,
    flexGrow: 0,
    borderColor: Color.ACCENT_BLUE_100,
  },
  overdueBookContainer: {
    borderColor: Color.ACCENT_ORANGE_100,
  },
});
