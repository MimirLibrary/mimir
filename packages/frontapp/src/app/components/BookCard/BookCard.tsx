import React, { FC } from 'react';
import styled from '@emotion/styled';
import { colors, dimensions } from '@mimir/ui-kit';
import BookImage from '../BookImage';
import TitleBook from '../TitleBook';
import ShortBookDescription from '../ShortBookDescription';
import BookStatus from '../BookStatus';

const BookCardWrapper = styled.div`
  height: 20.5rem;
  width: 12rem;
  background-color: ${colors.bg_secondary};
  border-radius: ${dimensions.xs};
  padding: ${dimensions.base};
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const DescriptionWrapper = styled.div`
  display: flex;
  justify-content: start;
  flex-direction: column;
`;

const BookCard: FC = () => {
  return (
    <BookCardWrapper>
      <BookImage src={''} />
      <DescriptionWrapper>
        <TitleBook title="The Psychology of Money Second part..." />
        <ShortBookDescription description="Psychology / Morgan Housel"></ShortBookDescription>
        <BookStatus status="Return till: 26.04" />
      </DescriptionWrapper>
    </BookCardWrapper>
  );
};

export default BookCard;
