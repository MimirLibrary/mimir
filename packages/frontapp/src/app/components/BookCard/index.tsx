import { FC } from 'react';
import styled from '@emotion/styled';
import { colors, dimensions } from '@mimir/ui-kit';
import bookImage from '../../../assets/MOC-data/BookImage.png';
import { useNavigate } from 'react-router-dom';
import { DateTime } from '@mimir/global-types';
import BookStatus from '../BookStatus';
export interface IBookCardProps {
  src?: string | null;
  title?: string;
  date?: DateTime;
  status?: string;
  author?: string;
  category?: string;
  id?: string;
}

const BookCardWrapper = styled.div`
  flex-shrink: 0;
  width: 12rem;
  background-color: ${colors.bg_secondary};
  border-radius: ${dimensions.xs};
  padding: ${dimensions.base};
  display: flex;
  flex-direction: column;
  cursor: pointer;
  transition: box-shadow 0.3s;

  :hover {
    box-shadow: 0px 6px 14px -6px rgba(24, 39, 75, 0.08),
      0px 10px 32px -4px rgba(24, 39, 75, 0.08);
  }
  @media (min-width: ${dimensions.phone_width}) {
    min-height: 19.5rem;
  }

  @media (max-width: ${dimensions.phone_width}) {
    flex-direction: row;
    height: min-content;
    width: 100%;
  }
`;

const DescriptionWrapper = styled.div`
  display: flex;
  justify-content: start;
  flex-direction: column;
`;

const BookImage = styled.img`
  width: 7rem;
  height: 12rem;
  margin-bottom: ${dimensions.xs_2};
  align-self: center;
  @media (max-width: ${dimensions.phone_width}) {
    width: 5rem;
    height: 8rem;
    margin-right: ${dimensions.base};
  }
`;

const TitleBook = styled.h3`
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  max-width: 10rem;
  font-weight: 500;
  font-size: ${dimensions.sm};
  line-height: ${dimensions.base};
  color: ${colors.main_black};
`;

const DescriptionBook = styled.p`
  max-width: 10rem;
  font-weight: 300;
  font-size: ${dimensions.sm};
  line-height: ${dimensions.base};
  color: ${colors.description_gray};
  margin: ${dimensions.xs_2} 0;
`;

const BookCard: FC<IBookCardProps> = ({
  src = '',
  title = '',
  author = '',
  status,
  category,
  date,
  id,
}) => {
  const navigate = useNavigate();
  const handleItemRedirect = () => {
    navigate(`/item/${id}`);
  };

  return (
    <BookCardWrapper onClick={handleItemRedirect} data-testid="bookCard">
      <BookImage src={src || bookImage} data-testid="bookImage" />
      <DescriptionWrapper>
        <TitleBook data-testid="bookTitle">{title}</TitleBook>
        <DescriptionBook>{category + ' / ' + author}</DescriptionBook>
        <BookStatus status={status} date={date} />
      </DescriptionWrapper>
    </BookCardWrapper>
  );
};

export default BookCard;
