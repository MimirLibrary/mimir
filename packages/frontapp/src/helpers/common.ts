export const shortenText = (text: string, numberToShorten: number) => {
  return text.trim().slice(0, numberToShorten);
};
