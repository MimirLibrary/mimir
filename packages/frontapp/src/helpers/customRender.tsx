import '@testing-library/jest-dom';
import { render, RenderOptions } from '@testing-library/react';
import React, { FC, ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { ApolloProvider } from '@apollo/client';
import { client } from '@mimir/apollo-client';

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string | null) => str,
      i18n: {
        changeLanguage: () =>
          new Promise(() => {
            return;
          }),
      },
    };
  },
}));

jest.unmock('react-redux');

const wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <BrowserRouter>
      <ApolloProvider client={client}>
        <Provider store={store}>{children}</Provider>
      </ApolloProvider>
    </BrowserRouter>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => {
  return render(ui, { wrapper, ...options });
};

export { customRender as render };
