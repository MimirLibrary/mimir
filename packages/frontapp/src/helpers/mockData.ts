export const mockStatus = [
  {
    person_id: 1,
    status: 'Free',
    created_at: '2022-09-08T03:58:39.228Z',
    id: 1,
    material_id: 1,
  },
];

export const mockMaterial = [
  {
    __typename: 'Material',
    author: 'test Author',
    category: 'test Category',
    created_at: '2022-09-08T03:58:39.228Z',
    description: 'Test Description',
    id: '1',
    id_type: 'Test Id_Type',
    identifier: '1111111111',
    is_donated: false,
    notification: [],
    picture: '',
    statuses: mockStatus,
    title: 'test',
    type: 'Book',
    updated_at: '2022-09-08T03:58:39.228Z',
  },
];
