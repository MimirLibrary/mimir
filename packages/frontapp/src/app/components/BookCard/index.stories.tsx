import { Story, Meta } from '@storybook/react';
import BookCard, { IBookCardProps } from './index';

export default {
  component: BookCard,
  title: 'BookCard',
} as Meta;

const Template: Story<IBookCardProps> = (args: IBookCardProps) => (
  <BookCard {...args} />
);

export const Primary = Template.bind({});
Primary.args = {
  title: 'Harry Potter',
};
