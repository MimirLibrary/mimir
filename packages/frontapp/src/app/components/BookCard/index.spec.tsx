import { screen } from '@testing-library/react';
import { render } from '../../../helpers/customRender';
import bookImage from '../../../assets/MOC-data/BookImage.png';
import BookCard from './index';
import { useTranslation } from 'react-i18next';

describe('BookCard Component', () => {
  const { t } = useTranslation();

  it('should render correctly', () => {
    render(<BookCard src={bookImage} />);
    expect(screen.getByTestId('bookCard')).toBeInTheDocument();
  });

  it('always render Book Status', () => {
    render(<BookCard src={bookImage} status="Free" />);
    expect(screen.getByTestId('bookStatus')).toBeInTheDocument();
    expect(screen.getByTestId('bookStatus')).toHaveTextContent(
      t('Statuses.Free')
    );
  });

  it('if the picture is not given, use the default one', () => {
    render(<BookCard />);
    expect(screen.getByTestId('bookImage')).toHaveAttribute('src', bookImage);
  });

  it('renders correctly, with snapshots', () => {
    render(
      <BookCard
        src={bookImage}
        title="The test book"
        status="Free"
        author="Test"
        category="Test"
        id="31"
      />
    );
    expect(screen.getByTestId('bookCard')).toMatchSnapshot();
  });
});
