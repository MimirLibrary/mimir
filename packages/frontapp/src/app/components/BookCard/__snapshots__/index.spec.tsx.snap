// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCard Component renders correctly, with snapshots 1`] = `
.emotion-0 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 12rem;
  background-color: #FFFFFF;
  border-radius: 0.75rem;
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
  -webkit-transition: box-shadow 0.3s;
  transition: box-shadow 0.3s;
}

.emotion-0:hover {
  box-shadow: 0px 6px 14px -6px rgba(24, 39, 75, 0.08),0px 10px 32px -4px rgba(24, 39, 75, 0.08);
}

@media (min-width: 30rem) {
  .emotion-0 {
    min-height: 19.5rem;
  }
}

@media (max-width: 30rem) {
  .emotion-0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    height: -webkit-min-content;
    height: -moz-min-content;
    height: min-content;
    width: 100%;
  }
}

.emotion-1 {
  width: 7rem;
  height: 12rem;
  margin-bottom: 0.5rem;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

@media (max-width: 30rem) {
  .emotion-1 {
    width: 5rem;
    height: 8rem;
    margin-right: 1rem;
  }
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: start;
  justify-content: start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-3 {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  max-width: 10rem;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1rem;
  color: #333333;
}

.emotion-4 {
  max-width: 10rem;
  font-weight: 300;
  font-size: 0.875rem;
  line-height: 1rem;
  color: #4F4F4F;
  margin: 0.5rem 0;
}

@media (max-width: 30rem) {
  .emotion-5 {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
    padding: 0 4px;
    border-radius: 4px;
  }
}

.emotion-6 {
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1rem;
  color: #2AA816;
}

@media (max-width: 30rem) {
  .emotion-6 {
    font-size: 0.625rem;
  }
}

<div
  class="emotion-0"
  data-testid="bookCard"
>
  <img
    class="emotion-1"
    data-testid="bookImage"
    src="BookImage.png"
  />
  <div
    class="emotion-2"
  >
    <h3
      class="emotion-3"
      data-testid="bookTitle"
    >
      The test book
    </h3>
    <p
      class="emotion-4"
    >
      Test / Test
    </p>
    <div
      class="emotion-5"
      data-testid="bookStatus"
    >
      <p
        class="emotion-6"
      >
        Statuses.Free
      </p>
    </div>
  </div>
</div>
`;
